### ISSUE TITLE
JUnit 4.13 jar from gradle distribution should not leak into test runtime classpath  

### ISSUE THREAD

**Issue by breskeby** [gradle/gradle#23995]
### Expected Behavior

The junit library shipped from the gradle distribution (e.g. `gradle-8.0.1-all/aro4hu1c3oeioove7l0i4i14o/gradle-8.0.1/lib/junit-4.13.2.jar`) should not be exposed on the test runtime classpath 

### Current Behavior

With our own elasticsearch test infrastructure jar hello check we see the junit jar exposed on the classpath:

```
java.lang.RuntimeException: found jar hell in test classpath
	at org.elasticsearch.bootstrap.BootstrapForTesting.<clinit>(BootstrapForTesting.java:106)
	at org.elasticsearch.test.ESTestCase.<clinit>(ESTestCase.java:264)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:495)
	at java.base/java.lang.Class.forName(Class.java:474)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$2.run(RandomizedRunner.java:631)
Caused by: java.lang.IllegalStateException: jar hell!
class: junit.textui.TestRunner
jar1: /Users/rene/.gradle/caches/modules-2/files-2.1/junit/junit/4.12/2973d150c0dc1fefe998f834810d68f278ea58ec/junit-4.12.jar
jar2: /Users/rene/.gradle/wrapper/dists/gradle-8.0.1-all/aro4hu1c3oeioove7l0i4i14o/gradle-8.0.1/lib/junit-4.13.2.jar
	at org.elasticsearch.jdk.JarHell.checkClass(JarHell.java:315)
	at org.elasticsearch.jdk.JarHell.checkJarHell(JarHell.java:233)
	at org.elasticsearch.jdk.JarHell.checkJarHell(JarHell.java:84)
	at org.elasticsearch.bootstrap.BootstrapForTesting.<clinit>(BootstrapForTesting.java:104)
	... 5 more
```

### Context

This blocks us from updating to 8.0.1

### Steps to Reproduce 

1. checkout this branch of my elasticsearch repo: https://github.com/breskeby/elasticsearch/tree/update-gradle-80
2. run `./gradlew :server:test`

### Your Environment

This bug seems not environment specific

Build scan URL: https://gradle.com/s/mt47vvnf7j7pm

**Comment by eskatos** [comment]
Build scan doesn't show the junit jar from the distribution
https://scans.gradle.com/s/mt47vvnf7j7pm/dependencies?dependencies=junit&expandAll&focusedDependencyView=versions

This means that this jar is added to the test runtime classpath out of the dependency resolution. Maybe leaking in the test worker classpath.

**Comment by eskatos** [comment]
Thank you for providing a valid reproducer.

The issue is in the backlog of the relevant team and is prioritized by them.

**Comment by jvandort** [comment]
This change in behavior was likely added here: https://github.com/gradle/gradle/pull/22505/files#diff-64c5b6a7bf54fcf434aefc2031f95ca25daaa641f770764aca2baf3f559f4256L85-R122

We had some initial detection of duplicate jars, but we deemed the solution fragile and removed that functionality here: https://github.com/gradle/gradle/pull/22583

Note that we intentionally add our distribution jar on the classpath _last_ so that any user-provided junit jars are first on the classpath. 

I will talk with the team to see how we want to handle this, however for immediate unblockers, consider:
1. Upgrade to `junit-4.13.2` by not adding junit to the classpath yourself and instead using the automatically-added distribution-provided jar
2. Special-case your check to ignore these duplicate jars
3. Migrate to test suites where the version of the automatically-added junit dependency can be controlled.

**Comment by jjohannes** [comment]
I _think_ we also face an issue like this with JUnit5. Our own `junit-platform-engine` Jar is signed (don't ask why this is 😄 ).

Moving from Gradle `7.x` to `8.0.1` we get a `signer information does not match signer information of other classes in the same package` for `EngineDiscoveryListner`. This class does not yet exist in the JUnit version we are currently using. So I suspect this causing the issue as some classes are loaded from our own Jar and others from the distribution Jar.

This is just as another datapoint. We can solve this somehow here. 

But I think "having two overlapping Jars on the classpath and relying on Classpath ordering" is in general a dangerous setup. If I read the code correctly, you also do not do this for the _module path_ (because that always gives an error).

Why do you need to load the Jar from the distribution at all? The engine is a transitive runtime dependency of the API, so it should always be there anyway right? What do I miss? (Or maybe I misunderstand the the whole issue, it's kind of complex...)

**Comment by jvandort** [comment]
> But I think "having two overlapping Jars on the classpath and relying on Classpath ordering" is in general a dangerous setup. If I read the code correctly, you also do not do this for the _module path_ (because that always gives an error).

I agree the overlapping classes on the classpath is not a good setup to have. 

> Why do you need to load the Jar from the distribution at all? The engine is a transitive runtime dependency of the API, so it should always be there anyway right? What do I miss? (Or maybe I misunderstand the the whole issue, it's kind of complex...)

We load `junit-platform-launcher` (and engine + commons) from the distribution, which is not a transitive dependency of `junit-jupiter`. It is only a version dependency constraint, so without an explicit dependency it will not be pulled in. 

> In this context, I also don't really understand this suggestion:
> 
> > Upgrade to junit-4.13.2 by not adding junit to the classpath yourself and instead using the automatically-added distribution-provided jar
> 
> This is not a normal setup, right? I need to do something like `implementation("junit:junit")` to be able to write my test (have the JUnit4 API at compile time). If I remove it, I can't compile my tests anymore.

I was wrong here and you are correct. You would still need to include `junit` as a `compileOnly` dependency. 

---

In general, we expected this classpath duplication to happen so in this case will not be classifying it as a _regression_. Given that there are feasible workarounds to this issue, will be targeting these fixes for `8.1` in order to avoid holding up important bug fixes in `8.0.2`.  That's not to say this is not an important issue, and it is currently my primary focus. We are targeting `8.1` branching for friday, so the RC will follow relatively soon after. 

> Why do you need to load the Jar from the distribution at all? 

Keywords:
`application classpath`: The `testRuntimeClasspath`
`implementation classpath`: Gradle's code which configures junit. This effectively extends from the `application classpath`. After junit is configured, the test process launches the actual tests in the context of the `application classpath`, without the implementation classes present. 

This is a complex issue but mostly stems from backwards compatibility. These testing libraries have historically been loaded from the distribution for junit 4 and `junit-platform-launcher` onto the implementation classpath, however due to #22333 we needed to ensure the same version of the launcher was present on both the the application and implementation classpath. Our solution was to add the dependency to the application classpath, which "leaks" into the implementation classpath, thus satisfying our constraint. However, that same solution is causing duplicate classes in some cases. Since this constraint is not required for junit 4, we can solve the initial issue by only adding junit 4 to the implementation classpath and not including it the application classpath. 

This boils down to two cases:
* For junit 4, we can go back to old behavior of loading the junit jar to the implementation classpath only. We can do this since we don't have the requirement of the junit jar being the same on the application and implementation classpath. The reason we cannot avoid loading it altogether is _junit 3 support_. We use the distribution junit 4 jar to "host" the junit 3 testing infrastructure. If the user has junit 3 as an _implementation_ dependency, but does not provide a junit 4 jar themselves, historically this has worked. We cannot break this. 
* For junit platform, we _need_ to have the same version of the launcher on both the implementation and application classpaths. In order to support cases where the jars are loaded from the distribution (see below for more info), we chose this solution. 

In general, we should avoid the distribution loading. For user-defined test suites, this already does not happen. For user-defined `Test` tasks, and for the default test suite when no test framework is explicitly configured (via the suite), we have no control over the configurations and need to use distribution loading. This is the sad truth of not using test suites. 

To resolve this for junit 5, we will do the following:
* In order to _verifiably_ determine the user has not included junit-platform-launcher themselves, we need to actually go and load the application classpath and attempt to load relevant classes. This is very expensive, since we do not actually load the classpath in the Gradle daemon where the classpath is constructed -- only in the test process. 
* So, we will do some "ugly" jar string comparisons to quickly determine if the user has provided the launcher in the "obvious" way. 
* Tentatively, depending on performance implications, if we cannot find the launcher jar, we will construct an actual ClassLoader and attempt to load junit launcher jars _in the daemon_. This is expensive but is a necessary evil if the launcher classes are provided in a non-standard named jar, like a fat jar. 
* If we determine that the launcher jars are _not_ present on the application classpath, we will emit a deprecation warning and load from the distribution. Then in 9.0 we can remove the loading behavior entirely by either expecting the user to provide these jars on the test runtime classpath manually OR using test suites which does all this automatically. 

All of this is avoided with test suites, since the suite is "in control" of the test `implementation` and `runtimeOnly` dependencies, and thus is able to add dependencies there instead of relying on loading these jars from the distribution. However, when using the default test suite, and for "standalone", non-suite `Test` tasks, historic behavior was to load these jars from the distribution. For standalone tasks, the `Test` is not aware of the configurations it is getting its dependencies from, and thus must find another way to get its dependencies (the distribution). For the default `test` task, which is backed by a test suite, we still cannot influence the test configurations _by default_, for backwards compatibility reasons. Using `testing.suites.test.useJunitJupiter()` or `testing.suites.test.useJUnit()` overrides this by disabling the distribution-loading and instead adding the dependencies to the correct configurations. 

Sorry for the wall of text. Please let me know if you have further questions or ideas. I am open to any suggestions for solutions here, since I realllly don't like the "read the test classpath jars and see if the junit-platform-launcher-x.x.jar is there" approach.

**Comment by jvandort** [comment]
The fix for junit4 has made it into `8.0.2`. The junit-platform changes are more involved and are tentatively targeted for `8.1`

**Comment by jjohannes** [comment]
Thank you for the detailed explanation @jvandort and for taking the time to improve this.

**Comment by jvandort** [comment]
The fix for junit-platform will make it into `8.1`. A merge has been triggered: https://github.com/gradle/gradle/pull/24115
